Realm/ObjectServerTests/RealmServer.swift

```diff
@@ -0,0 +1,780 @@
+                    guard let adminProfile = adminProfile else {
+                        if let error = error {
+                            outError = error
+                        } else {
+                            outError = URLError(.badServerResponse)
+                        }
+                        return
+                    }
+                    
+                    adminSession = AdminSession(accessToken: accessToken, groupId: adminProfile.roles[0].groupId)
+                }
+            } catch {
+                outError = error
+            }
+        }.resume()
+        guard case .success = group.wait(timeout: .now() + 5) else {
+            outError = URLError(.cannotFindHost)
+            return (adminSession, outError)
+        }
+        return (adminSession, outError)
+    }
+}
+
+// MARK: RealmServer
+
+/**
+ A sandboxed server. This singleton launches and maintains all server processes
+ and allows for app creation.
+ */
+@objc(RealmServer)
+public class RealmServer : NSObject {
+    /// Shared RealmServer. This class only needs to be initialized and torn down once per test suite run.
+    @objc static let shared: RealmServer = RealmServer()
+    /// Process that runs the local mongo server. Should be terminated on exit.
+    private let mongoProcess = Process()
+    /// Process that runs the local backend server. Should be terminated on exit.
+    private let serverProcess = Process()
+    
+    /// The root URL of the project.
+    private lazy var rootUrl = URL(string: #file)!
+        .deletingLastPathComponent()
+        .deletingLastPathComponent()
+        .deletingLastPathComponent()
+    
+    /// The directory where mongo binaries and backing files are kept.
+    private lazy var mongoUrl = rootUrl
+        .appendingPathComponent("build")
+        .appendingPathComponent("mongodb-macos-x86_64-4.4.0-rc5")
+    
+    /// The directory where mongo stores its files. This is a unique value so that
+    /// we have a fresh mongo each run.
+    private lazy var mongoDataDirectory = ObjectId.generate().stringValue
+    
+    /// Whether or not this is a parent or child process.
+    private lazy var isParentProcess = (getenv("RLMProcessIsChild") == nil)
+    
+    /// The current admin session
+    private var session: Admin.AdminSession? = nil
+    
+    private override init() {
+        super.init()
+
+        if isParentProcess {
+            atexit {
+                _ = RealmServer.shared.tearDown
+            }
+
+            launchMongoProcess()
+            launchServerProcess()
+            let (adminSession, loginError) = Admin().login()
+            guard let session = adminSession else {
+                XCTFail("Could not initiate admin session: \(loginError?.localizedDescription ?? "unknown cause")")
+                return
+            }
+            self.session = session
+        }
+    }
+
+    /// Lazy teardown for exit only.
+    private lazy var tearDown: () = {
```

Comment:
Why is this a lazy var and not just a function?

Reply:
To act as a `dispatch_once` replacement.